### PR TITLE
Implement const files for RGBASM

### LINKED ISSUE #606: Add constants to sym file as well
What makes the SYM file useful is that it's not easy to know where a given label is located; however, not all constants are obvious, as in the following example.
```asm
SECTION "Gfx", ROMX
Gfx:
INCBIN "gfx.2bpp"

GFX_SIZE equ @ - Gfx ; What value do I have?
    PRINTT "{GFX_SIZE}\n" ; This can help, but it's easy to lose it in a busy `make` invocation
```
For compatibility with existing implementations, maybe output them in comments, at the end of the file, with a special comments preceding them?

Note that this ties to #483.

### PR DESCRIPTION
**Pull request by Rangi42**
Fixes #606

### CODE REVIEW

**Comment by vulcandth** [comment]
Doing as you said in the issue and padding the values to 8; would help users understand these are 32bit values from a glance. It may also help align the two columns. Downside is an increased file size, but I don't think that's much of a real issue. I think I like non-padded, but we may want to go with what will cause less confusion.

**Comment by Rangi42** [comment]
I think having the common cases of one or two-byte values being their natural widths is the least confusing / easiest to read.

**Comment by ISSOtm** [comment]
I'm a bit unconvinced, because *each* assembly file has to emit its own set of constants, and now downstream consumers will have to look for a *tree* of files (`.sym` lacks `include`).

I'd rather like the constants to be emitted to the object file (unconditionally?), and then RGBLINK emits them to the same sym file as usual.

Though then that begs the question of how to resolve conflicts (identical name but different values), which generating a per-TU sym file only punts to the consumer (increasing the potential for fragmentation, which I *really* don't like).

Noise can probably reduced by dropping variables? But it's still possible for constants to conflict. Maybe print none if they conflict, instead print a comment listing the various values.

**Comment by Rangi42** [comment]
I think the use cases for logging labels and constants are very different, and the complexity + edge cases of your suggestion here show why we should go with your original idea of an rgbasm flag.

We already have a .sym file spec that doesn't allow for bankless symbols. I think putting constants, say, after the labels with an artificial 00 bank, would be confusing.

I expect some tools parse the .sym file to understand all the labels across a game, and mixing constants in there could corrupt their results. (Meaning "I know I have some scripts like that lying around, and emulators might not deal with it well either.")

Labels are fundamentally a link-time, whole-program entity, but constants are fundamentally an artifact of one single pass of rgbasm over continuous source. Examples:

- Generic variables like x/n/i, or pret's `const_value` (similar to `_RS`) that get used in FOR loops and don't really matter, but should not be provoking warnings when they end up with different values in different TUs (let alone errors).
- Values that are meaningfully different in different TUs. How to indicate them both in one file? Even something like `42 foo ; path/to/source.asm` might be insufficient if rgbasm flags like `-D` influenced its value (I'm sure *someone* has a reused source file with different flags).

Basically, my expected use case for this PR feature is for small single-file projects; or large projects that output for one file during debugging; or large projects that build a special constants_only.asm file if they want to log the constants they use across the whole project.

**Comment by ISSOtm** [comment]
> I think the use cases for logging labels and constants are very different, and the complexity + edge cases of your suggestion here show why we should go with your original idea of an rgbasm flag.
> 
> We already have a .sym file spec that doesn't allow for bankless symbols. I think putting constants, say, after the labels with an artificial 00 bank, would be confusing.

The sym file spec [supports bank-less symbols](https://rgbds.gbdev.io/sym/#location).

> I expect some tools parse the .sym file to understand all the labels across a game, and mixing constants in there could corrupt their results. (Meaning "I know I have some scripts like that lying around, and emulators might not deal with it well either.")

If those don't conform to the spec, well, then either it's a problem with the scripts, or the spec is worthless. I'd prefer enforcing it, otherwise we're back to square one wrt. the problem that spec was written to resolve.

> Outputting all constants in the object file would make exporting them kind of pointless, and bloat the object for a rare use case.

Exporting would still have the “unique across entire program” and “usable from other TUs” semantics. I agree about the file size; I am reminded of why C compilers have `-g` opt-in.

> Labels are fundamentally a link-time, whole-program entity, but constants are fundamentally an artifact of one single pass of rgbasm over continuous source. Examples:

Counter-example: labels defined as constants, possibly because several banks are being used as “overlays” of sorts, or because `UNION` is inconvenient. (I've had at least two separate instances of that, CRC-self-test code and ACE payloads.)

> * Generic variables like x/n/i, or pret's `const_value` (similar to `_RS`) that get used in FOR loops and don't really matter, but should not be provoking warnings when they end up with different values in different TUs (let alone errors).

Those would be covered by the “only emit constants, not variables” heuristic.

> * Values that are meaningfully different in different TUs. How to indicate them both in one file? Even something like `42 foo ; path/to/source.asm` might be insufficient if rgbasm flags like `-D` influenced its value (I'm sure _someone_ has a reused source file with different flags).

The closest-matching example I can think of is if someone were hardcoding banks. (Cross-bank `INCBIN`, for example.)

As I mentioned, those cases could be resolved by emitting

```
; 42 foo (in 42.asm)
; 69 foo (in 69.asm)
```

or

```
42 foo ; (in 42.asm)
; 69 foo (in 69.asm)
```

> Basically, my expected use case for this PR feature is for small single-file projects; or large projects that output for one file during debugging; or large projects that build a special constants_only.asm file if they want to log the constants they use across the whole project.

Then wouldn't something akin to `gmake -P` be more appropriate?

**Comment by Rangi42** [comment]
@ISSOtm Following discussion on Discord, I've renamed this to a "dump file", and only output the `EQU` constants to it (not `=` variables). Its format is technically a valid .sym file (as you pointed out, a location [can be](https://rgbds.gbdev.io/sym/#location) a bankless number), but we're not bound to keep it that way, and in future could output variables, strings, charmaps, more metadata, etc.

**Comment by aaaaaa123456789** [comment]
You can't output constants meaningfully as a symfile-equivalent file because constants can be larger than addresses. `DEF FOOBAR EQU $deadbeef` is valid. If the symfile spec supports that, it's an oversight — that's not a valid address at all.

**Comment by Rangi42** [comment]
@ISSOtm I'm not entirely happy with the name "dump file". Considering some other options:

- `make -p` prints a "data base" file of the rules and variable values
- `gcc -d` prints a "debugging dump"; in particular `-dM` prints the `#define` values
- `nasm -l` prints a "listing" file; in particular, `-Ls` prints all single-line macro definitions (i.e. `equ` values)

**Comment by aaaaaa123456789** [comment]
If it's a file with the value of constants, why not call it a constants file?

**Comment by Rangi42** [comment]
> If it's a file with the value of constants, why not call it a constants file?

Because "we might want to add more info to it later than just constants".

**Comment by aaaaaa123456789** [comment]
Cross the bridge when you get to the water, maybe? Renaming the file is cheap; that's just documentation, not an API.

**Comment by vulcandth** [comment]
I like the idea of going with `listing` as Sylvie mentioned. Makes sense to me for the outputs we are talking about. (Constants, but perhaps other stuff too going forward.)

**Comment by aaaaaa123456789** [comment]
A listing is a file containing source code. That's what the word has meant for decades.

**Comment by Rangi42** [comment]
Not in ma'am with `-Ls`!

**Comment by aaaaaa123456789** [comment]
The listing file in NASM _does_ contain the assembly code.

**Comment by Rangi42** [comment]
Now renamed to "state file".

File: src/asm/output.cpp [hunk 1]
```diff
@@ -347,10 +351,57 @@ void out_WriteObject() {
...
+		file = fopen(stateFileName.c_str(), "w");
+	} else {
+		stateFileName = "<stdout>";
+		file = fdopen(STDOUT_FILENO, "w");
+	}
+	if (!file)
+		err("Failed to open state file '%s'", stateFileName.c_str());
+	Defer closeFile{[&] { fclose(file); }};
+
+	// Only write the non-builtin numeric constants
```
Comment (ISSOtm) [inline src/asm/output.cpp:369]: I think it would be more useful to write more user-defined symbols, such as `EQUS`. To that effect, employing RGBASM's syntax sounds like a good idea (reusing the provided syntax highlighting), and we could use a `.dump.asm` extension in the shell completions.

Perhaps we can add an extra CLI flag that lets the user decide what to request in the dump file; I'd consider using the same syntax as rgbobj for that.

**Review by ISSOtm — CHANGES_REQUESTED** [review]

**Comment by ISSOtm** [comment]
> Cross the bridge when you get to the water, maybe? Renaming the file is cheap; that's just documentation, not an API.

The CLI *does*, however, qualify as an API.

**Comment by aaaaaa123456789** [comment]
> > Cross the bridge when you get to the water, maybe? Renaming the file is cheap; that's just documentation, not an API.
> 
> The CLI _does_, however, qualify as an API.

True, but adding a new long option (and leaving the old one as legacy) is essentially free.

**Comment by ISSOtm** [comment]
The current code architecture precludes adding “just” a long option, we have to burn a short option for each. Also, I don't want to add legacy, because we already have enough of that as-is; hence pushing to get this design right (enough) on first impl.

**Comment by Rangi42** [comment]
It sounds like we still have a lot of possibilities to consider for what a solution to this issue looks like (sym-file syntax, or something more flexible? also output charmaps? also output `EQUS`? anything else we might want to support? what to even call it?) so I'm closing this PR.
